### PR TITLE
Add support to jdl blueprints property.

### PR DESCRIPTION
**Pull request by mshima**
Implement environment builder.
Unify the environment cli environment operations into one builder.

Fixes https://github.com/jhipster/generator-jhipster/issues/11815

-   Please make sure the below checklist is followed for Pull Requests.

-   [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [ ] Tests are added where necessary
-   [ ] Documentation is added/updated where necessary
-   [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by mshima** [comment]
@MathieuAA can you review this one?

**Comment by MathieuAA** [comment]
@mshima of course

File: cli/environment-builder.js [hunk 1]
```diff
@@ -0,0 +1,299 @@
+const assert = require('assert');
```
Comment (MathieuAA) [inline cli/environment-builder.js:1]: the copyright is missing

File: test/cli/environment-builder.spec.js [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+            beforeEach(() => {
+                blueprintsWithVersion = envBuilder.getBlueprintsWithVersion();
+            });
+
+            it('returns an empty object', () => {
+                expect(blueprintsWithVersion).to.deep.equal({});
+            });
+        });
+
+        describe('when blueprints was passed by command', () => {
```
Comment (MathieuAA) [inline test/cli/environment-builder.spec.js:74]: `when blueprints was passed by command` => `when blueprints were passed by command`

File: test/cli/environment-builder.spec.js [hunk 3]
```diff
@@ -0,0 +1,273 @@
+/* eslint-disable no-unused-expressions */
```
Comment (MathieuAA) [inline test/cli/environment-builder.spec.js:1]: no copyright here too

File: test/cli/run-yeoman-process.spec.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
+/* eslint-disable no-unused-expressions */
```
Comment (MathieuAA) [inline test/cli/run-yeoman-process.spec.js:1]: no copyright here too

File: test/templates/blueprint-cli-shared/cli/sharedOptions.js [hunk 5]
```diff
@@ -1,3 +1,4 @@
 module.exports = {
```
Comment (MathieuAA) [inline test/templates/blueprint-cli-shared/cli/sharedOptions.js:1]: no copyright here too

File: test/templates/blueprint-cli-shared/generators/bar/index.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+const Generator = require('yeoman-generator');
```
Comment (MathieuAA) [inline test/templates/blueprint-cli-shared/generators/bar/index.js:1]: no copyright here too

File: test/templates/blueprint-cli-shared/generators/bar/index.js [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+
+module.exports = class extends Generator {
+    initializing() {
+        /* eslint-disable no-console */
```
Comment (MathieuAA) [inline test/templates/blueprint-cli-shared/generators/bar/index.js:5]: there's no console statement, you can remove it

File: test/templates/blueprint-cli-shared/generators/bar/index.js [hunk 8]
```diff
@@ -0,0 +1,7 @@
+const Generator = require('yeoman-generator');
+
+module.exports = class extends Generator {
+    initializing() {
```
Comment (MathieuAA) [inline test/templates/blueprint-cli-shared/generators/bar/index.js:4]: does it actually do something by default? is that why you've overridden it?

File: cli/environment-builder.js [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+const chalk = require('chalk');
+const _ = require('lodash');
+const path = require('path');
+const Environment = require('yeoman-environment');
+const { CLI_NAME, logger } = require('./utils');
+const { normalizeBlueprintName, packageNameToNamespace, loadYoRc, loadBlueprintsFromConfiguration } = require('../generators/utils');
+
+module.exports = class EnvironmentBuilder {
```
Comment (MathieuAA) [inline cli/environment-builder.js:9]: can you add a comment saying what a JHipster environment is? i.e. standard yo env + blueprints
Maybe this class should be renamed JHipsterEnvironmentBuilder, or something

File: cli/environment-builder.js [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+        this.blueprintsWithVersion = this._getAllBlueprintsWithVersion();
+        return this;
+    }
+
+    /**
+     * Get blueprints with version.
+     *
+     * return {Object} blueprints with versions.
+     */
+    getBlueprintsWithVersion() {
```
Comment (MathieuAA) [inline cli/environment-builder.js:77]: it isn't used anywhere, can we remove it?

File: cli/environment-builder.js [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+        });
+        return this;
+    }
+
+    /**
+     * Load blueprints from argv, .yo-rc.json.
+     *
+     * @return {EnvironmentBuilder} this for chaining.
+     */
+    loadBlueprints() {
```
Comment (MathieuAA) [inline cli/environment-builder.js:67]: why not private, it's only used here

File: cli/environment-builder.js [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+    constructor(env) {
+        this.env = env;
+    }
+
+    /**
+     * Lookup current jhipster generators.
+     *
+     * @return {EnvironmentBuilder} this for chaining.
+     */
+    lookupJHipster() {
```
Comment (MathieuAA) [inline cli/environment-builder.js:50]: why not private, it's only used here

File: cli/environment-builder.js [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+    getBlueprintsWithVersion() {
+        return this.blueprintsWithVersion;
+    }
+
+    /**
+     * Lookup current loaded blueprints.
+     *
+     * @return {EnvironmentBuilder} this for chaining.
+     */
+    lookupBlueprints(options) {
```
Comment (MathieuAA) [inline cli/environment-builder.js:86]: why not private, it's only used here

File: cli/environment-builder.js [hunk 14]
```diff
@@ -0,0 +1,299 @@
...
+        }
+        return this;
+    }
+
+    /**
+     * Load sharedOptions from jhipster and blueprints.
+     *
+     * @return {EnvironmentBuilder} this for chaining.
+     */
+    loadSharedOptions() {
```
Comment (MathieuAA) [inline cli/environment-builder.js:100]: why not private, it's only used here

File: test/templates/blueprint-cli-shared/generators/bar/index.js (hunk 8 above)
Comment (mshima) [reply test/templates/blueprint-cli-shared/generators/bar/index.js:4]: generators cannot be empty.

File: cli/environment-builder.js (hunk 9 above)
Comment (mshima) [reply cli/environment-builder.js:9]: I had problems before putting jsdoc in the class, they were duplicating.
I had to move to the constructor to fix that.

**Comment by MathieuAA** [comment]
I'll let you merge it when you want, great work :)

**Comment by mshima** [comment]
@MathieuAA thanks.

**Pull request merged**